File: www/ymchat.js
var exec = require("cordova/exec");

const YmChatAPI = {
  setBotId: (botId) => {
    console.log("setting bot id");
    exec(null, null, "ymchat", "setBotId", [botId]);
  },
  setDeviceToken: (token) => {
    exec(null, null, "ymchat", "setDeviceToken", [token]);
  },
  setEnableSpeech: (speech) => {
    exec(null, null, "ymchat", "setEnableSpeech", [speech]);
  },
  setAuthenticationToken: (token) => {
    exec(null, null, "ymchat", "setAuthenticationToken", [token]);
  },
  showCloseButton: (closeBot) => {
    exec(null, null, "ymchat", "showCloseButton", [closeBot]);
  },
  setCustomURL: (url) => {
    exec(null, null, "ymchat", "setCustomURL", [url]);
  },
  setPayload: (payload) => {
    exec(null, null, "ymchat", "setPayload", [payload]);
  },
  onEventFromBot: (eventListener) => {
    exec(eventListener, null, "ymchat", "onEventFromBot", []);
  },
  onBotClose: (onBotCloseEvent) => {
    exec(onBotCloseEvent, null, "ymchat", "onBotClose", []);
  },
  startChatbot: (success, failure) => {
    exec(success, failure, "ymchat", "startChatbot", []);
  },
  closeBot: () => {
    exec(null, null, "ymchat", "closeBot", []);
  },
  unlinkDeviceToken: (apiKey, success, failure) => {
    exec(success, failure, "ymchat", "unlinkDeviceToken", [apiKey])
  },
  registerDevice: (apiKey, success, failure) => {
    exec(success, failure, "ymchat", "registerDevice", [apiKey])
  },
  getUnreadMessagesCount: (success, failure) => {
    exec(success, failure, "ymchat", "getUnreadMessagesCount", [])
  },
  setVersion: (version) => {
    exec(null, null, "ymchat", "setVersion", [version])
  },
  setCustomLoaderURL: (url) => {
    exec(null, null, "ymchat", "setCustomLoaderURL", [url]);
  },
  setStatusBarColor: (color) => {
    exec(null, null, "ymchat", "setStatusBarColor", [color]);
  },
  setCloseButtonColor: (color) => {
    exec(null, null, "ymchat", "setCloseButtonColor", [color]);
  },
  setDisableActionsOnLoad: (shouldDisableActionsOnLoad) => {
    exec(null, null, "ymchat", "setDisableActionsOnLoad", [shouldDisableActionsOnLoad]);
  },
  useLiteVersion: (shouldUseLiteVersion) => {
    exec(null, null, "ymchat", "useLiteVersion", [shouldUseLiteVersion]);
  },
  reloadBot: () => {
    exec(null, null, "ymchat", "reloadBot", []);
  },
  setMicIconColor: (color) => {
    exec(null, null, "ymchat", "setMicIconColor", [color]);
  },
  setMicBackgroundColor: (color) => {
    exec(null, null, "ymchat", "setMicBackgroundColor", [color]);
  },
  useSecureYmAuth: (shouldUseSecureYmAuth) => {
    exec(null, null, "ymchat", "useSecureYmAuth", [shouldUseSecureYmAuth]);
  },
  setThemeBotName: (name) => {
    exec(null, null, "ymchat", "setThemeBotName", [name]);
  },
  setThemeBotDescription: (description) => {
    exec(null, null, "ymchat", "setThemeBotDescription", [description]);
  },
  setThemePrimaryColor: (color) => {
    exec(null, null, "ymchat", "setThemePrimaryColor", [color]);
  },
  setThemeSecondaryColor: (color) => {
    exec(null, null, "ymchat", "setThemeSecondaryColor", [color]);
  },
  setThemeBotBubbleBackgroundColor: (color) => {
    exec(null, null, "ymchat", "setThemeBotBubbleBackgroundColor", [color]);
  },
  setThemeBotIcon: (iconUrl) => {
    exec(null, null, "ymchat", "setThemeBotIcon", [iconUrl]);
  },
  setThemeBotClickIcon: (iconUrl) => {
    exec(null, null, "ymchat", "setThemeBotClickIcon", [iconUrl]);
  },
  setChatContainerTheme: (theme) => {
    exec(null, null, "ymchat", "setChatContainerTheme", [theme]);
  },
  revalidateToken: (token, refreshSession) => {
    exec(null, null, "ymchat", "revalidateToken", [token, refreshSession]);
  },
  sendEventToBot: (code, data) => {
    exec(null, null, "ymchat", "sendEventToBot", [code, data]);
  }
};

module.exports = YmChatAPI;

File: src/ios/ymchat.m
/********* ymchat.m Cordova Plugin Implementation *******/

#import <Cordova/CDV.h>

@import YMChat;

@interface ymchat : CDVPlugin<YMChatDelegate> 
{
    CDVInvokedUrlCommand* onEvent;
    CDVInvokedUrlCommand* onBotClosed;
}

@end

@implementation ymchat
- (void)setBotId:(CDVInvokedUrlCommand*)command
{
    NSString* botId = [command.arguments objectAtIndex:0];
    YMConfig *config = [[YMConfig alloc] initWithBotId:botId];
    YMChat.shared.config = config;
    YMChat.shared.enableLogging = true;
}

- (void)setDeviceToken:(CDVInvokedUrlCommand*)command
{
    NSString* deviceToken = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.deviceToken = deviceToken;
}

- (void)setEnableSpeech:(CDVInvokedUrlCommand*)command
{
    BOOL enableSpeech = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.enableSpeech = enableSpeech;
}

- (void)setAuthenticationToken:(CDVInvokedUrlCommand*)command
{
    NSString* authToken = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.ymAuthenticationToken = authToken;
}

- (void)showCloseButton:(CDVInvokedUrlCommand*)command
{
    BOOL closeBot = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.showCloseButton = closeBot;
}

- (void)setCustomURL:(CDVInvokedUrlCommand*)command
{
    NSString* url = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.customBaseUrl = url;
}

- (void)setPayload:(CDVInvokedUrlCommand*)command
{
    NSDictionary* payload = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.payload = payload;
}

- (void)showCloseBot:(CDVInvokedUrlCommand*)command
{
    BOOL enableSpeech = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.enableSpeech = enableSpeech;
}

- (void)onEventFromBot:(CDVInvokedUrlCommand*)command
{
    onEvent = command;
}

- (void)onBotClose:(CDVInvokedUrlCommand*)command
{
    onBotClosed = command;
}

- (void)startChatbot:(CDVInvokedUrlCommand*)command
{
    assert(YMChat.shared.config != nil);
    YMChat.shared.delegate = self;
    [[YMChat shared] startChatbotWithAnimated:YES error: nil completion:nil];
}

- (void)closeBot:(CDVInvokedUrlCommand*)command
{
    [[YMChat shared] closeBot];
}

- (void)reloadBot:(CDVInvokedUrlCommand*)command
{
    [[YMChat shared] reloadBotAndReturnError:nil];
}

- (void)onEventFromBotWithResponse:(YMBotEventResponse *)response {
    CDVPluginResult* pluginResult = nil;
    NSMutableDictionary *event = [[NSMutableDictionary alloc]initWithCapacity:10];
    [event setObject:response.code forKey:@"code"];
    if(response.data){
        [event setObject:response.data forKey:@"data"];
    }
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:event];
    [pluginResult setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:onEvent.callbackId];
}

- (void) onBotClose {
    CDVPluginResult* pluginResult = nil;
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Hello"];
    [pluginResult setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:onBotClosed.callbackId];
}

- (void)setVersion:(CDVInvokedUrlCommand*)command
{
    assert(YMChat.shared.config != nil);
    NSNumber* version = [command.arguments objectAtIndex:0];
    YMChat.shared.config.version = version.integerValue;
}

- (void) unlinkDeviceToken:(CDVInvokedUrlCommand*)command {
    if(YMChat.shared.config) {
        NSString* apiKey = [command.arguments objectAtIndex:0];
        [[YMChat shared] unlinkDeviceTokenWithApiKey:apiKey ymConfig:YMChat.shared.config success:^{
            CDVPluginResult* pluginResult = nil;
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } failure:^(NSString * _Nonnull failureMessage) {
            CDVPluginResult* pluginResult = nil;
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:failureMessage];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    } else {
        CDVPluginResult* pluginResult = nil;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Bot id not found"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void) registerDevice:(CDVInvokedUrlCommand*)command {
    NSString* apiKey = [command.arguments objectAtIndex:0];
    if(YMChat.shared.config) {
        [[YMChat shared] registerDeviceWithApiKey:apiKey ymConfig:YMChat.shared.config success:^{
            CDVPluginResult* pluginResult = nil;
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } failure:^(NSString * _Nonnull failureMessage) {
            CDVPluginResult* pluginResult = nil;
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:failureMessage];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    } else {
        CDVPluginResult* pluginResult = nil;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Bot id not found"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void) getUnreadMessagesCount:(CDVInvokedUrlCommand*)command {
    if(YMChat.shared.config) {
        [[YMChat shared] getUnreadMessagesCountWithYmConfig:YMChat.shared.config success:^(NSString * _Nonnull unreadCount) {
            CDVPluginResult* pluginResult = nil;
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:unreadCount];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } failure:^(NSString * _Nonnull failureMessage) {
            CDVPluginResult* pluginResult = nil;
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:failureMessage];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    } else {
        CDVPluginResult* pluginResult = nil;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Bot id not found"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void)setCustomLoaderURL:(CDVInvokedUrlCommand*)command
{
    NSString* url = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.customLoaderUrl = url;
}

- (void)setStatusBarColor:(CDVInvokedUrlCommand*)command
{
    NSString* color = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.statusBarColor = [self getColorFromHexString:color];
}

- (void)setCloseButtonColor:(CDVInvokedUrlCommand*)command
{
    NSString* color = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.closeButtonColor = [self getColorFromHexString:color];
}

- (void)setDisableActionsOnLoad:(CDVInvokedUrlCommand*)command
{
    BOOL shouldDisableActionsOnLoad = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.disableActionsOnLoad = shouldDisableActionsOnLoad;
}

- (void)useLiteVersion:(CDVInvokedUrlCommand*)command
{
    BOOL shouldUseLiteVersion = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.useLiteVersion = shouldUseLiteVersion;
}

- (void)setMicIconColor:(CDVInvokedUrlCommand*)command
{
    NSString* color = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.enableSpeechConfig.fabIconColor = [self getColorFromHexString:color];
}

- (void)setMicBackgroundColor:(CDVInvokedUrlCommand*)command
{
    NSString* color = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.enableSpeechConfig.fabBackgroundColor = [self getColorFromHexString:color];
}

- (void)useSecureYmAuth:(CDVInvokedUrlCommand*)command
{
    BOOL shouldUseSecureYmAuth = [command.arguments objectAtIndex:0];
    assert(YMChat.shared.config != nil);
    YMChat.shared.config.useSecureYmAuth = shouldUseSecureYmAuth;
}

- (void)revalidateToken:(CDVInvokedUrlCommand*)command
{
    NSString* token = [command.arguments objectAtIndex:0];
    BOOL refreshSession = [command.arguments objectAtIndex:1];
    [[YMChat shared] revalidateTokenWithToken:token refreshSession:refreshSession error:nil];
}

- (void)sendEventToBot:(CDVInvokedUrlCommand*)command
{
    NSString* code = [command.arguments objectAtIndex:0];
    NSDictionary* data = [command.arguments objectAtIndex:1];
    YMEventModel* event = [[YMEventModel alloc] initWithCode:code data:data];
    [[YMChat shared] sendEventToBotWithEvent:event error:nil];
}

- (UIColor *)getColorFromHexString:(NSString *)hexString {
    unsigned rgbValue = 0;
    NSScanner *scanner = [NSScanner scannerWithString:hexString];
    if ([(NSString*) hexString characterAtIndex:0] == '#') {
        [scanner setScanLocation:1]; // bypass '#' character
    } else {
        [scanner setScanLocation:0];
    }
    [scanner scanHexInt:&rgbValue];
    return [UIColor colorWithRed:((rgbValue & 0xFF0000) >> 16)/255.0 green:((rgbValue & 0xFF00) >> 8)/255.0 blue:(rgbValue & 0xFF)/255.0 alpha:1.0];
}

@end
